src/HttpHeader.h

```diff
@@ -118,8 +118,8 @@ class HttpHeader
     bool hasNamed(const SBuf &s, String *value = 0) const;
     /// \deprecated use SBuf method instead.
     bool hasNamed(const char *name, unsigned int namelen, String *value = 0) const;
-    String getByNameListMember(const char *name, const char *member, const char separator) const;
-    String getListMember(Http::HdrType id, const char *member, const char separator) const;
+    SBuf getByNameListMember(const char *name, const char *member, const char separator) const;
```

Comment:
Please do not stuff HTTP API conversion into this PR. Introducing those new APIs correctly is a large, difficult task. It is unnecessary in this PR. I do not want to end up spending days reviewing and/or rewriting those API changes as a part of this totally unrelated work. Upgrading String to SBuf alone is not a significant enough change to (legitimately) insist on the corresponding API overhaul.

Reply:
> Upgrading String to SBuf alone is not a significant enough change to (legitimately) insist on the corresponding API overhaul.

This does not compute. This PR *does* an API overhaul, but not to the apparently desirable API. I am asking for the overhaul changes to be left out of scope and the desirable API be added for use currently only by this new caller.

Reply:
> This PR _does_ an API overhaul

I assuming we are still talking about the HttpHeader API. No, this PR does not overhaul that API. It changes the return type of a few methods from one string type to another string type. That is not an API overhaul. That is a small/minor optimization. This string type replacement is not a sufficient ground to insist on adding some new API methods.

Reply:
@Alex, the internals of these getByNameListMember and getListMember functions has been completely re-written using new data types and new logic sequences. So there is no less review needed for the current approach. In fact the opposite..

What I am disliking is that by re-using the old member names this PR adds a review of all the old callers to the workload. If this PR were only shuffling the code around, or using the old member signatures - no problem. But a full re-write of the member both internally and symbol types is significant change. Either we agree to keep the "Foo get*(...)" style of members, or we push the switch to your "bool has*()" design whenever anyone touches the class methods (as done here).

Reply:
> getListMember functions has been completely re-written using new data types and new logic sequences.

It may seem so but is not the reality. I see one 'logic' change: the new getListMember() got additional safety array-boundary check `static_cast<size_t>(ilen) > keyLen`. Other logic is the same. 

> review of all the old callers

If you'd examined these callers, you probably noticed that:
* some of the old callers naturally required SBuf though (format/Format.cc), using the clumsy StringToSBuf conversion. The new code there (which should be in line with the similar calls) would require these conversions either, which is not perfect.
* some of the old callers only cared about the emptiness of the obtained string. Is it really important what method we call there, String::size() or SBuf::isEmpty()?

Reply:
> So there is no less review needed for the current approach. In fact the opposite..

That is just your opinion about the time _I_ will have to spend. In reality, the effort I usually spend on reviewing or writing APIs is often far greater than the effort spent on reviewing method implementations, especially basic ones. And I already reviewed those implementations, so we are comparing a significant API writing effort with zero.

> Either we agree to keep the "Foo get*(...)" style of members, or we push the switch to your "bool has*()" design whenever anyone touches the class methods (as done here).

You have no right to impose such rigid restrictions and false dichotomies.

If you insist on rewriting these APIs in this PR, then I suggest reverting the corresponding upgrade to SBuf (since you are abusing that upgrade to demand out-of-scope API changes). I hope you do not insist, especially after this PR fixed so many bugs, but I was also hoping you would not even _start_ this conversation, so I am prepared to be proven wrong.

@yadij, do you insist on rewriting these APIs in this PR?

Reply:
I am sorry that you went and put a lot of work into a review that IMO should have only gone to the overview stage before opting to do the smaller code changes. Resolving this issue now rather than getting into yet another idealogical discussion.

Having the API use SBuf is a good long=term result anyway, it just means that this approach of changing so many callers implicit behaviour makes it to wide-spread to risk backporting for v4 stable series. Which is unfortunate given the PROXYv2 issues this fixes.

Reply:
> Having the API use SBuf [...] means that this approach of changing so many callers implicit behaviour makes it to wide-spread to risk backporting for v4 stable series.

If you are worried about any callers that did not get explicitly changed, then it is easy to expose all of them during a v4 port -- a temporary rename of the affected functions will automatically expose all callers in a full build. I do that sort of checks a lot, and they are easy to apply and revert. If you want Factory to perform this check for your v4 port, just ask. We can also prepare a v4-specific PR.